prov/verbs/src/verbs_info.c

```diff
@@ -1397,6 +1397,25 @@ static int fi_ibv_handle_ib_ud_addr(const char *node, const char *service,
 	return ret;
 }
 
+static void fi_ibv_remove_nosrc_info(struct fi_info **info)
+{
+	struct fi_info **fi = info, *next;
+	while (*fi && ((*fi)->ep_attr->type == FI_EP_MSG)) {
+		if (!(*fi)->src_addr) {
+			VERBS_INFO(FI_LOG_FABRIC, "Not reporting fi_info "
+				   "corresponding to domain: %s as it has no IP"
+				   "address configured\n",
+				   (*fi)->domain_attr->name);
+			next = (*fi)->next;
+			(*fi)->next = NULL;
+			fi_freeinfo(*fi);
+			*fi = next;
+		} else {
+			fi = &(*fi)->next;
+		}
+	}
+}
```

Comment:
I'm getting lost in pointer-pointers.  But this function doesn't actually modify 'info', correct?  Basically, what happens if the first entry or all entries is removed?

Reply:
Yes, it doesn't modify info. It only modifies what it points to. When first entry is removed, line 1412 would replace the first entry with the second. If all entries are removed *fi and hence *info would be set to next which would be NULL (since it's the next of last entry).

This follows the following template to remove an element from a singly linked list without using additional head, previous pointers:
```
while (*node_dptr) {
	if (remove(*node_dptr)) {
		next = (*node_dptr)->next;
		free(*node_dptr);
		*node_dptr = next;
	} else {
		node_dptr = &((*node_dptr)->next);
	}
}
```